lib/view_component/test_helpers.rb

```diff
@@ -244,5 +249,9 @@ def __vc_test_helpers_preview_class
     rescue NameError
       raise NameError, "`render_preview` expected to find #{result}, but it does not exist."
     end
+
+    def __vc_render_preview_controller
+      @vc_render_preview_controller ||= __vc_test_helpers_build_controller(Rails.application.config.view_component.preview_controller.constantize)
```

Comment:
This is slightly changing behavior since we're now memoizing this value. Is this okay to memoize or do we need to reset it each render? I _think_ we want to keep the old behavior and have it re-evaluate per-call to `render_preview` so each call has a fresh controller instance?

Reply:
Added a `#teardown` method that clears the variable.

Reply:
I mentioned it above, but not every test needs a preview controller. Can we revert this back to the previous, lazy loaded implementation to avoid extra work in test suites?

Reply:
Unfortunately to support devise I don't see this being possible. We can't predict when `render_preview` is going to be used. The entire problem is that to support devise, setup needs to happen before `render_preview` is ever called. If we don't memoize, then the test may still work in most cases, but keep in mind it wont be accurate, as you're warden request environ would be a different instance than your render_preview request environ.

One possible alternative that might meet your requirements is a new concern, which sets up a test environment for devise and `render_preview`. That way a user is actively opting in essentially stating "I am using this pattern", and then update documentation to the concern's existence.

Honestly if it weren't for the fact that this project already has a bunch of minitests built before I started working on it, I'd just move the whole thing over to rspec and be done with it :D

Reply:
Something like

```ruby
module ViewComponent::TestHelpers::DevisePreviewHelper
   def self.included(base)
      base.instance_eval do
        setup do
          @request = __vc_render_preview_controller.request
        end
         
        teardown do
          @request = nil
          @vc_render_preview_controller = nil
        end
          
      # ....
   end
    
    def __vc_render_preview_controller
      @vc_render_preview_controller ||= super
    end
      
    
end
```

Reply:
To help clarify things, are devise helpers only broken for previews or for previews + render_inline?

Reply:
Just checked, it works for `render_inline`. I believe it is because `render_preview` uses an actual controller instance, causing it to pick up Warden's hooks.